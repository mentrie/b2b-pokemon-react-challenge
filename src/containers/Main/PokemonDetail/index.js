import React, { Component } from 'react';

class PokemonDetails extends Component {

  render() {
    return (
      <div>
        {this.props.match}
        <p onClick={this.props.backToHome}> Back </p>
        Details
      </div>
    );
  }
};

export default PokemonDetails;
