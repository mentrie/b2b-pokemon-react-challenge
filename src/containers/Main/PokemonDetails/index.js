import React, { Component } from 'react';
import { apiRequest } from '../../../api';

class PokemonDetails extends Component {
  state = {
    details: null
  }

  async componentWillMount() {
    if (this.state.details == null) {
      const { match } = this.props;
    //   try {
    //     const result = await apiRequest('/')
    //   }
    }
  }
  render() {
    return (
      <div>
        <p onClick={this.props.backToHome}> Back </p>
        Details
      </div>
    );
  }
};

export default PokemonDetails;
