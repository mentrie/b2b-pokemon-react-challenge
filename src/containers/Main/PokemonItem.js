import React from 'react';
import PropTypes from 'prop-types';
import Link from 'react-router-dom/Link';
import { getPokemanId, pokemonImageUrl, formatName } from '../../helpers/utils';

const PokemonItem = ({ pokemon, onSelect }) => {
  const id = getPokemanId(pokemon.url);

  return (
    <Link to={`/${pokemon.name}/${id}`} onClick={onSelect} className="main__pokemon__item">
      <img src={pokemonImageUrl(id)} alt={pokemon.name} />
    </Link>
  );
}

PokemonItem.propTypes = {
  pokemon: PropTypes.shape({
    id: PropTypes.number,
    url: PropTypes.string
  })
}

export default PokemonItem;