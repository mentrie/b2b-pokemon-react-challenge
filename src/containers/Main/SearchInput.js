import React from 'react';

const SearchInput = ({ value, onChange }) => {
  return (
    <input
      type="text"
      placeholder="Search for PokÃ©mon here"
      value={value}
      onChange={onChange}
    />
  );
};

export default SearchInput;
