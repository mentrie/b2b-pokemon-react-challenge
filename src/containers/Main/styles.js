export default {
  overlay: {
    position: 'fixed',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    backgroundColor: 'rgba(255, 255, 255, 0.87)'
  },
  content: {
    position: 'absolute',
    top: 0,
    left: 0,
    right: 0,
    bottom: 0,
    border: '.1rem solid #ccc',
    background: '#3c3c3c',
    overflow: 'auto',
    WebkitOverflowScrolling: 'touch',
    borderRadius: '.3rem',
    outline: 'none',
    padding: '.5rem'
  }
};
