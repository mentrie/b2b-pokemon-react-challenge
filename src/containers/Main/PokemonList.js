import React from 'react';
import PropTypes from 'prop-types';
import PokemonItem from './PokemonItem';

const PokemonList = ({ pokemons, onSelect }) => {
  const renderPokemons = () => {
    return pokemons.map(pokemon => (
      <PokemonItem key={pokemon.name}
        pokemon={pokemon}
        onSelect={onSelect}
      />
    ));
  };

  return (
    <div className="main__pokemon__wrapper">
      {renderPokemons()}
    </div>
  );
};

PokemonList.propTypes = {
  pokemons: PropTypes.arrayOf(PropTypes.shape({
    name: PropTypes.string,
    url: PropTypes.string.isRequired
  }))
};

export default PokemonList;
