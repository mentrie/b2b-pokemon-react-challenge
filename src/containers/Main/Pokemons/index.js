import React, { Component } from 'react';
import DetailModal from 'react-modal';
import withRouter from 'react-router-dom/withRouter';
import Nprogress from 'nprogress';
import ReactPlaceholder from 'react-placeholder';
import 'nprogress/nprogress.css';
import 'react-placeholder/lib/reactPlaceholder.css';
import { apiRequest } from '../../../api';
import { PAGE_SIZE } from '../../../constants';
import PokemonList from './PokemonList';
import SearchInput from '../../../components/SearchInput';
import Pagination from '../../../components/Pagination';

export class Pokemons extends Component {
  state = {
    pokemons: [],
    error: '',
    search: '',
    currentPage: 1
  }

  async componentWillMount() {
    DetailModal.setAppElement('body');
    this.props.history.replace('/');

    if (this.state.pokemons.length === 0) {
      Nprogress.start();
      try {
        const { results } = await apiRequest('/?limit=50');
        this.setState({pokemons: results});
      } catch (e) {
        this.setState({error: 'Something went wrong'});
      }
      Nprogress.done();
    }
  }

  handleSearch = ({target: {value}}) => this.setState(({search}) => ({search: value}));

  onSelectPage = (currentPage) => this.setState({currentPage});

  getDisplayedData() {
    const { search, pokemons } = this.state;
    const searchQuery = new RegExp(search, 'gi');

    return search.length
      ? pokemons.filter(pokemon => pokemon.name.search(searchQuery) >= 0)
      : pokemons;
  }

  render() {
    const { pokemons, search, currentPage } = this.state;
    const displayPokemons = this.getDisplayedData();
    const totalPage = Math.ceil(displayPokemons.length / PAGE_SIZE) || 0;
    const pokemonsToRender = displayPokemons.slice(Math.max(0, (currentPage - 1) * PAGE_SIZE), currentPage * PAGE_SIZE);

    return (
      <div className="pokemon">
        <ReactPlaceholder showLoadingAnimation type="media" rows={7} ready={pokemons.length > 0}>
          <SearchInput onChange={this.handleSearch} value={search} />
          {pokemonsToRender && pokemonsToRender.length
            ? <PokemonList pokemons={pokemonsToRender} onSelect={this.props.handleSelect} />
            : <div className="main__not-found">No Pokemon found :)</div>
          }
        </ReactPlaceholder>
        <Pagination
          current={currentPage}
          total={totalPage}
          onNext={() => this.onSelectPage(Math.min(currentPage + 1, totalPage))}
          onPrev={() => this.onSelectPage(Math.max(1, currentPage - 1))}
        />
      </div>
    );
  }
}

export default withRouter(Pokemons);
