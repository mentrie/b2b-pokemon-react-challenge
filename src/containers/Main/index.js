import React, { Component } from 'react'
import Nprogress from 'nprogress';
import ReactPlaceholder from 'react-placeholder';
import 'nprogress/nprogress.css';
import 'react-placeholder/lib/reactPlaceholder.css';
import { apiRequest } from '../../api';
import { PAGE_SIZE } from '../../constants';
import PokemonList from './PokemonList';
import SearchInput from './SearchInput';

export class Main extends Component {
  state = {
    pokemons: [],
    error: '',
    search: ''
  }

  handleSearch = ({ target: {value}}) => {
    this.setState({search: value})
  }

  filterPokemon(pokemons) {
    // todo
  }

  async componentWillMount() {

    if (this.state.pokemons.length === 0) {
      Nprogress.start();
      try {
        const { results } = await apiRequest('/?limit=50');
        this.setState({pokemons: results});
      } catch (e) {
        this.setState({error: 'Something went wrong'})
      }
      Nprogress.done();
    }
  }

  render() {
    const { pokemons, search } = this.state;
    const displayPokemons = pokemons.slice(0, PAGE_SIZE)
    return (
      <main className="main">
        <SearchInput onChange={this.handleSearch} value={search}/>
        <ReactPlaceholder showLoadingAnimation type="media" rows={7} ready={pokemons.length > 0}>
          {displayPokemons && displayPokemons.length? <PokemonList pokemons={displayPokemons} />: <div>No Pokemon found</div>}
        </ReactPlaceholder>
      </main>
    )
  }
}

export default Main;
