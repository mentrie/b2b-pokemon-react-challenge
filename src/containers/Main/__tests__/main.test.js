import React from 'react';
import Router from 'react-router-dom/Router';
import createHistory from 'history/createBrowserHistory';
import Main from '../index';

test('renders component', () => {
  const component = shallow(<Main />);

  expect(component).toMatchSnapshot();
});

test('renders subcomponents', () => {
  const component = shallow(<Main />);
  
  expect(component.find('Pokemons').length).toBeDefined();
  expect(component.find('PokemonDetails').length).toBeDefined(); 
  expect(component.find('DetailModal').length).toBeDefined();     
});

test('renders with history props', () => {
  const history = createHistory();
  const component = <Router history={history}><Main /></Router>;

  expect(component).toMatchSnapshot();
  expect(component.props.history).toBeDefined()
});



