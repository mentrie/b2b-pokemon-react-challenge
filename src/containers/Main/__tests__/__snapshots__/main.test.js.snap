// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders component 1`] = `
<Route
  render={[Function]}
/>
`;

exports[`renders with history props 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <withRouter(Main) />
</Router>
`;
